packages/vite/src/node/plugins/html.ts

```diff
@@ -138,83 +131,134 @@ export const isAsyncScriptMap = new WeakMap<
+): node is DefaultTreeAdapterMap['element'] {
+  return node.nodeName[0] !== '#'
+}
+
+function traverseNodes(
+  node: DefaultTreeAdapterMap['node'],
+  visitor: (node: DefaultTreeAdapterMap['node']) => void
+) {
+  visitor(node)
+  if (
+    nodeIsElement(node) ||
+    node.nodeName === '#document' ||
+    node.nodeName === '#document-fragment'
+  ) {
+    node.childNodes.forEach((childNode) => traverseNodes(childNode, visitor))
+  }
+}
+
 export async function traverseHtml(
   html: string,
   filePath: string,
-  visitor: NodeTransform
+  visitor: (node: DefaultTreeAdapterMap['node']) => void
 ): Promise<void> {
   // lazy load compiler
-  const { parse, transform } = await import('@vue/compiler-dom')
-  // @vue/compiler-core doesn't like lowercase doctypes
-  html = html.replace(/<!doctype\s/i, '<!DOCTYPE ')
-  try {
-    const ast = parse(html, { comments: true })
-    transform(ast, {
-      nodeTransforms: [visitor]
-    })
-  } catch (e) {
-    handleParseError(e, html, filePath)
-  }
+  const { parse } = await import('parse5')
+  const ast = parse(html, {
+    sourceCodeLocationInfo: true,
+    onParseError: (e: ParserError) => {
+      handleParseError(e, html, filePath)
+    }
+  })
+  traverseNodes(ast, visitor)
 }
 
-export function getScriptInfo(node: ElementNode): {
-  src: AttributeNode | undefined
+export function getScriptInfo(node: DefaultTreeAdapterMap['element']): {
+  src: Token.Attribute | undefined
+  sourceCodeLocation: Token.Location | undefined
   isModule: boolean
   isAsync: boolean
 } {
-  let src: AttributeNode | undefined
+  let src: Token.Attribute | undefined
+  let sourceCodeLocation: Token.Location | undefined
   let isModule = false
   let isAsync = false
-  for (let i = 0; i < node.props.length; i++) {
-    const p = node.props[i]
-    if (p.type === NodeTypes.ATTRIBUTE) {
-      if (p.name === 'src') {
-        src = p
-      } else if (p.name === 'type' && p.value && p.value.content === 'module') {
-        isModule = true
-      } else if (p.name === 'async') {
-        isAsync = true
-      }
+  for (let i = 0; i < node.attrs.length; i++) {
+    const p = node.attrs[i]
+    if (p.name === 'src') {
+      src = p
+      sourceCodeLocation = node.sourceCodeLocation?.attrs!['src']
+    } else if (p.name === 'type' && p.value && p.value === 'module') {
+      isModule = true
+    } else if (p.name === 'async') {
+      isAsync = true
     }
   }
-  return { src, isModule, isAsync }
+  return { src, sourceCodeLocation, isModule, isAsync }
+}
+
+const attrValueWithQuotesStartRE = /=[\s\t\n\r]*"/
+
+export function overwriteAttrValue(
+  s: MagicString,
+  sourceCodeLocation: Token.Location,
+  newValue: string
+): MagicString {
+  const srcString = s.slice(
+    sourceCodeLocation!.startOffset,
+    sourceCodeLocation!.endOffset
+  )
+  const valueWithQuotes = srcString.match(attrValueWithQuotesStartRE)
+  const valueOffset =
+    1 + (valueWithQuotes ? srcString.indexOf('"') : srcString.indexOf('='))
```

Comment:
Are single quotes also allowed in this context?

Reply:
I think yes, but perhaps the easier way I think is to replace the entire key+value instead, since technically the value could have no quotes, and if the `newValue` has a space in between it could bring problems.

Replacing the entire thing should be fine too as we don't generate sourcemaps for html.

Reply:
Oh, good catch. I wanted to avoid changing the spacing, even if we don't generate sourcemaps. If the user has `attr = "foo"`, I didn't want to convert it to `attr="foo"`. Same now about `attr='foo'` to `attr="foo"`. I think that we should at least respect the choice of `'` or `"` here. IIUC, this always worked in this way in Vite, spacing was respected. Would it be ok to keep doing this?

Reply:
Yeah I think it would be nice if we can keep it that way too as long as it doesn't make the code too complex. I don't think it would be a big issue though if we decide to always normalize to using `"` so we can avoid potential edge cases like `attr=foo` (no quotes). As we could simply `JSON.stringify` and call it a day 😬 

(Sorry for the late reply)

Reply:
Done in [vitejs/vite@`471d229` (#9678)](https://github.com/vitejs/vite/pull/9678/commits/471d2299ef5137c8d625c43e57f50fbf2dbd3262), it now supports `"`, `'`, and no wrapped values respecting spacing for all of them.